### PR TITLE
More as completed methods

### PR DESCRIPTION
**Pull request by jcrist**
Add two new methods to `AsCompleted`:

- `is_empty`: returns True if there no waiting futures, False otherwise
- `update`: a bulk version of `add`

I'm not set on the names. Since we have `add`, I used `update` instead of `extend` to better mirror a `set`. For `is_empty`, the `Queue` class has an `empty` method, but I don't like that name as it could be confused with `clear`.

I also thought about implementing `__bool__` as `not self.is_empty()`, but didn't yet. Could go either way on that.

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -2656,9 +2668,8 @@ def __aiter__(self):
         return self
 
     def __next__(self):
-        with self.lock:
-            if not self.futures and self.queue.empty():
-                raise StopIteration()
+        if self.is_empty():
```
Comment (jcrist) [inline distributed/client.py]: May also be able to replace the similar check in `__anext__`. That one doesn't use a lock though, and I'm not sure if that was intentional or not.

File: distributed/client.py [hunk 2]
```diff
@@ -2638,16 +2638,28 @@ def track_future(self, future):
                 self.queue.put_nowait(future)
             self.condition.notify()
 
+    def update(self, futures):
+        """ Add multiple futures to the collection.
+
+        The added futures will emit from the iterator once they finish"""
+        for f in futures:
+            with self.lock:
```
Comment (mrocklin) [inline distributed/client.py]: Should we swap the order of the loop and the lock here?

File: distributed/client.py (hunk 1 above)
Comment (mrocklin) [inline distributed/client.py]: Yeah, the async code doesn't need locks as long as we're consistent between yield/await points.

File: distributed/tests/test_as_completed.py [hunk 3]
```diff
@@ -65,6 +65,22 @@ def test_as_completed_add(loop):
             assert total == expected
 
 
+def test_as_completed_update(loop):
+    with cluster() as (s, [a, b]):
+        with Client(s['address'], loop=loop) as c:
+            total = 0
+            todo = list(range(10))
+            expected = sum(map(inc, todo))
+            ac = AsCompleted([])
```
Comment (mrocklin) [inline distributed/tests/test_as_completed.py]: FWIW we've been using the `as_completed` syntax everywhere.  Actually, I wouldn't mind renaming the class to `as_completed`.

File: distributed/client.py [hunk 4]
```diff
@@ -2638,16 +2638,28 @@ def track_future(self, future):
                 self.queue.put_nowait(future)
             self.condition.notify()
 
+    def update(self, futures):
```
Comment (pitrou) [inline distributed/client.py]: `__init__` can now call `update`, by the way.

File: distributed/client.py (hunk 2 above)
Comment (jcrist) [inline distributed/client.py]: Oop, that's what I meant to do. Fixed.

File: distributed/client.py (hunk 4 above)
Comment (jcrist) [inline distributed/client.py]: Thanks, fixed.

**Comment by mrocklin** [comment]
Merged.  Thanks @jcrist !

**Pull request merged**